Mathlib/MeasureTheory/SetSemiring.lean

```diff
@@ -0,0 +1,279 @@
+  (hC.diff_eq_Union' s hs t ht).choose \ {∅}
+
+lemma empty_not_mem_diffFinset (hC : IsSetSemiring C) (hs : s ∈ C) (ht : t ∈ C)
+    [DecidableEq (Set α)] :
+    ∅ ∉ hC.diffFinset hs ht := by
+  simp only [diffFinset, mem_sdiff, Finset.mem_singleton, eq_self_iff_true, not_true,
+    and_false_iff, not_false_iff]
+
+lemma diffFinset_subset (hC : IsSetSemiring C) (hs : s ∈ C) (ht : t ∈ C) [DecidableEq (Set α)] :
+    ↑(hC.diffFinset hs ht) ⊆ C := by
+  simp only [diffFinset, coe_sdiff, coe_singleton, diff_singleton_subset_iff]
+  exact (hC.diff_eq_Union' s hs t ht).choose_spec.1.trans (Set.subset_insert _ _)
+
+lemma pairwiseDisjoint_diffFinset (hC : IsSetSemiring C) (hs : s ∈ C) (ht : t ∈ C)
+    [DecidableEq (Set α)] :
+    PairwiseDisjoint (hC.diffFinset hs ht : Set (Set α)) id := by
+  simp only [diffFinset, coe_sdiff, coe_singleton]
+  exact Set.PairwiseDisjoint.subset (hC.diff_eq_Union' s hs t ht).choose_spec.2.1
+      (Set.diff_subset _ _)
+
+lemma sUnion_diffFinset (hC : IsSetSemiring C) (hs : s ∈ C) (ht : t ∈ C) [DecidableEq (Set α)] :
+    ⋃₀ hC.diffFinset hs ht = s \ t := by
+  rw [(hC.diff_eq_Union' s hs t ht).choose_spec.2.2]
+  simp only [diffFinset, coe_sdiff, coe_singleton, diff_singleton_subset_iff]
+  rw [sUnion_diff_singleton_empty]
+
+lemma not_mem_diffFinset (hC : IsSetSemiring C) (hs : s ∈ C) (ht : t ∈ C) [DecidableEq (Set α)] :
+    t ∉ hC.diffFinset hs ht := by
+  intro hs_mem
+  suffices t ⊆ s \ t by
+    have h := @disjoint_sdiff_self_right _ t s _
+    specialize h le_rfl this
+    simp only [Set.bot_eq_empty, Set.le_eq_subset, subset_empty_iff] at h
+    refine hC.empty_not_mem_diffFinset hs ht ?_
+    rwa [← h]
+  rw [← hC.sUnion_diffFinset hs ht]
+  exact subset_sUnion_of_mem hs_mem
+
+lemma sUnion_insert_diffFinset (hC : IsSetSemiring C) (hs : s ∈ C) (ht : t ∈ C) (hst : t ⊆ s)
+    [DecidableEq (Set α)] :
+    ⋃₀ insert t (hC.diffFinset hs ht) = s := by
+  conv_rhs => rw [← union_diff_cancel hst, ← hC.sUnion_diffFinset hs ht]
+  simp only [mem_coe, sUnion_insert]
+
+lemma disjoint_sUnion_diffFinset (hC : IsSetSemiring C) (hs : s ∈ C) (ht : t ∈ C)
+    [DecidableEq (Set α)] :
+    Disjoint t (⋃₀ hC.diffFinset hs ht) := by
+  rw [hC.sUnion_diffFinset]
+  exact disjoint_sdiff_right
+
+lemma pairwiseDisjoint_insert_diffFinset (hC : IsSetSemiring C) (hs : s ∈ C) (ht : t ∈ C)
+    [DecidableEq (Set α)] :
+    PairwiseDisjoint (insert t (hC.diffFinset hs ht) : Set (Set α)) id := by
+  have h := hC.pairwiseDisjoint_diffFinset hs ht
+  refine PairwiseDisjoint.insert_of_not_mem h (hC.not_mem_diffFinset hs ht) fun u hu ↦ ?_
+  simp_rw [id.def]
+  refine Disjoint.mono_right ?_ (hC.disjoint_sUnion_diffFinset hs ht)
+  simp only [Set.le_eq_subset]
+  exact subset_sUnion_of_mem hu
+
+end diffFinset
+
+section diffFinset₀
+
+variable {I : Finset (Set α)}
+
+/-- In a semiring of sets `C`, for all set `s ∈ C` and finite set of sets `I ⊆ C`, there is a
+finite set of sets in `C` such that `s \ ⋃₀ I = ⋃₀ (hC.diffFinset₀ hs I hI)`.
```

Comment:
```suggestion
finite set of sets in `C` whose union is `s \ ⋃₀ I`.
```
The sentence doesn't really make sense otherwise, as `hC.diffFinset₀ hs I hI` is not defined yet.